src/main/java/org/springframework/integration/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -372,89 +385,82 @@ public boolean isLongLived() {
+											}
+											highestFetchedOffset = Math.max(highestFetchedOffset,
+													kafkaMessage.getMetadata().getNextOffset());
 										}
-										highestFetchedOffset =
-												Math.max(highestFetchedOffset, kafkaMessage.getMetadata().getNextOffset());
-									}
-									fetchOffsets.replace(batch.getPartition(), highestFetchedOffset);
-									// if there are still messages on server, we can go on and retrieve more
-									if (highestFetchedOffset < batch.getHighWatermark()) {
-										partitionsWithRemainingData.add(batch.getPartition());
+										fetchOffsets.replace(batch.getPartition(), highestFetchedOffset);
 									}
 								}
+								// handle errors
+								if (result.getErrors().size() > 0) {
+									// find partitions with leader errors and
+									PartitionIterable<Map.Entry<Partition, Short>> partitionByLeaderErrors = partition(
+											result.getErrors().entrySet(), new IsLeaderErrorPredicate());
+									RichIterable<Partition> partitionsWithLeaderErrors = partitionByLeaderErrors
+											.getSelected().collect(keyFunction);
+									resetLeaders(partitionsWithLeaderErrors);
+
+									PartitionIterable<Map.Entry<Partition, Short>> partitionsWithOffsetsOutOfRange = partitionByLeaderErrors
+											.getRejected().partition(new IsOffsetOutOfRangePredicate());
+									resetOffsets(
+											partitionsWithOffsetsOutOfRange.getSelected().collect(keyFunction).toSet());
+									// it's not a leader issue
+									listenedPartitions.removeAllIterable(
+											partitionsWithOffsetsOutOfRange.getRejected().collect(keyFunction));
+								}
 							}
-							// handle errors
-							if (result.getErrors().size() > 0) {
-								hasErrors = true;
-
-								// find partitions with leader errors and
-								PartitionIterable<Map.Entry<Partition, Short>> partitionByLeaderErrors =
-										partition(result.getErrors().entrySet(), new IsLeaderErrorPredicate());
-								RichIterable<Partition> partitionsWithLeaderErrors =
-										partitionByLeaderErrors.getSelected().collect(keyFunction);
-								resetLeaders(partitionsWithLeaderErrors);
-
-								PartitionIterable<Map.Entry<Partition, Short>> partitionsWithOffsetsOutOfRange =
-										partitionByLeaderErrors.getRejected()
-												.partition(new IsOffsetOutOfRangePredicate());
-								resetOffsets(partitionsWithOffsetsOutOfRange.getSelected()
-										.collect(keyFunction)
-										.toSet());
-								// it's not a leader issue
-								stopFetchingFromPartitions(partitionsWithOffsetsOutOfRange.getRejected()
-										.collect(keyFunction));
+							else {
+								active = false;
 							}
 						}
 						catch (ConsumerException e) {
-							resetLeaders(fetchPartitions.toImmutable());
+							active = false;
+							// the connection is broken, terminate the task
+							kafkaTemplate.getConnectionFactory().disconnect(brokerAddress);
+							resetLeaders(listenedPartitions.toImmutable());
```

Comment:
Since we are going to leave the loop and `fetchTasksByBroker.remove(brokerAddress)`  in the `finally`, I think it would be safer to `resetLeaders` after `fetchTasksByBroker.remove()` there in the `finally`, rather than here and finish with the unreachable broker.